File: src/main/java/net/ndrei/teslacorelib/items/BaseTieredAddon.java
package net.ndrei.teslacorelib.items;

import net.minecraft.creativetab.CreativeTabs;
import net.ndrei.teslacorelib.tileentities.SidedTileEntity;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by CF on 2017-04-01.
 */
public class BaseTieredAddon extends BaseAddon {
    public BaseTieredAddon(String modId, CreativeTabs tab, String registryName) {
        super(modId, tab, registryName);
    }

    protected boolean hasSameFunction(BaseTieredAddon other) {
        return (other != null) &&
                (this.getClass().isAssignableFrom(other.getClass()) || (this.getAddonFunction().equals(other.getAddonFunction())));
    }

    protected String getAddonFunction() {
        return this.getRegistryName().toString();
    }

    protected int getTier() {
        return 1;
    }

    @Override
    public boolean canBeAddedTo(SidedTileEntity machine) {
        int tier = this.getTier();
        Map<Integer, BaseAddon> tiers = new HashMap<>();

        List<BaseAddon> addons = machine.getAddons();
        if (addons != null) {
            for (BaseAddon addon : addons) {
                if (!(addon instanceof BaseTieredAddon)) {
                    break;
                }
                BaseTieredAddon tiered = (BaseTieredAddon)addon;

                if (this.hasSameFunction(tiered)) {
                    if (tiered.getTier() == tier) {
                        // already has an addon with same tier and function
                        return false;
                    }

                    tiers.put(tiered.getTier(), tiered);
                }
            }
        }

        for(int index = 1; index < tier; index++) {
            if (!tiers.containsKey(index)) {
                // missing an addon with an inferior tier
                return false;
            }
        }
        return true;
    }
//
//    @Override
//    public void onRemoved(ItemStack addon, SidedTileEntity machine) {
//        super.onRemoved(addon, machine);
//
//        if (!ItemStackUtil.isEmpty(addon) && (addon.getItem() == this)) {
//            List<BaseAddon> addons = machine.getAddons();
//            if (addons != null) {
//                int tier = this.getTier();
//
//                for (BaseAddon ba : addons) {
//                    if (!(ba instanceof BaseTieredAddon)) {
//                        break;
//                    }
//                    BaseTieredAddon tiered = (BaseTieredAddon)ba;
//
//                    if (this.hasSameFunction(tiered)) {
//                        if (tiered.getTier() > tier) {
//                            // already has an addon with same tier and function
//                            machine.removeAddon(ba, true);
//                        }
//                    }
//                }
//            }
//        }
//    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/items/BaseTieredAddon.kt
package net.ndrei.teslacorelib.items

import net.minecraft.creativetab.CreativeTabs
import net.ndrei.teslacorelib.tileentities.SidedTileEntity
import java.util.*

/**
 * Created by CF on 2017-06-27.
 */
open class BaseTieredAddon(modId: String, tab: CreativeTabs, registryName: String)
    : BaseAddon(modId, tab, registryName) {
    protected fun hasSameFunction(other: BaseTieredAddon?): Boolean {
        return other != null && (this.javaClass.isAssignableFrom(other.javaClass) || this.addonFunction == other.addonFunction)
    }

    protected open val addonFunction: String
        get() = this.registryName!!.toString()

    protected open val tier: Int
        get() = 1

    fun isTierValid(machine: SidedTileEntity, tier: Int, ignoreSameTier: Boolean): Boolean {
        if (tier == 1 && ignoreSameTier) {
            return true
        }

        val tiers = HashMap<Int, BaseAddon>()

        val addons = machine.addons
        for (addon in addons) {
            if (addon !is BaseTieredAddon) {
                break
            }
            val tiered = addon

            if (this.hasSameFunction(tiered)) {
                if (tiered.tier == tier && !ignoreSameTier) {
                    // already has an addon with same tier and function
                    return false
                }

                tiers.put(tiered.tier, tiered)
            }
        }

        // missing an addon with an inferior tier
        return ((tier == 1) && (tiers.count() == 0)) || (1..tier - 1).any { tiers.containsKey(it) }
    }

    override fun canBeAddedTo(machine: SidedTileEntity): Boolean {
        return this.isTierValid(machine, this.tier, false)
    }

    override fun isValid(machine: SidedTileEntity): Boolean {
        return super.isValid(machine) && this.isTierValid(machine, this.tier, true)
    }
}
